recover.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"net/http"
+	"runtime"
+)
+
+// A recoverMux recovers from panics during an http.Handler, sending an error
+// response to a client in the event a panic occurs.
+type recoverMux struct {
+	mux *http.ServeMux
+
+	DumpStack bool // set to true to dump the stack to the client on a panic
+}
+
+// NewRecoverMux returns a new recoverMux, creating and wrapping a new
+// http.ServeMux.
+func newRecoverMux() *recoverMux {
+	return &recoverMux{
+		mux: http.NewServeMux(),
+	}
+}
+
+// HandleFunc wraps the caller's handler with a recovery handler which recovers
+// from panics in the caller's handler. Response data is only written if the
+// caller's handler completes without panicking.
+func (rmux *recoverMux) HandleFunc(pattern string, handler func(http.ResponseWriter, *http.Request)) {
+	f := func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if x := recover(); x != nil {
+				rmux.recoverHandler(w, r)
+			}
+		}()
+		rw := newResponseWriter(w)
```

Comment:
What's the point of wrapping the writer?

Reply:
The intent was to buffer the user's handler's response in the event that it writes part of its response but then panics before returning. In that case, the recovering handler can discard that response and write only the "Something went wrong" message to the client. I'll take a look at my commentary to ensure it describes that. I just implemented that with a bytes.Buffer. Is there a better, more idiomatic way to go about something like this?